# python_3_7

**Useful Links:**

## [python3 Google Doc](https://docs.google.com/document/d/1wKzv4vmRUpprhvyDLXwxIBHUIgOHxUKvZ8noK43Rfvs/edit?usp=sharing)

## [python3 standard library](https://docs.python.org/3/library)

---

[File Directories & recursion](https://www.pythoncentral.io/series/python-recursive-file-and-directory-manipulation/)

[GLOB](https://docs.python.org/3/library/glob.html#module-glob)

[Subprocess](https://docs.python.org/3/library/subprocess.html#module-subprocess)

[SHUTIL](https://docs.python.org/3/library/shutil.html#module-shutil)

---

# IMORTANT IDIOMS


**call MAIN function to run script(s)**

`# Standard boilerplate to call the main() function.
if __name__ == '__main__':
  main()`

**try, except, else statements**

`
errorMessage = str(sys.exc_info()[0]) + "\n"
errorMessage = errorMessage + str(sys.exc_info()[1]) + "\n"
errorMessage = errorMessage + str(sys.exc_info()[2]) + "\n"
exc_type, exc_obj, exc_tb = sys.exc_info()
fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
print("TYPE : " + str(exc_type) + "\n")
print("FILE : " + str(fname) + "\n")
print("LINE-NO : " + str(exc_tb.tb_lineno) + "\n")
print("MESSAGE : \n" + str(errorMessage))`

**ANOTHER POSSIBLE OUTPUT:**

`
errorMessage = str(sys.exc_info()[0]) + "n\t\t"
errorMessage = errorMessage + str(sys.exc_info()[1]) + "\n\t\t"
errorMessage = errorMessage + str(sys.exc_info()[2]) + "\n"
exc_type, exc_obj, exc_tb = sys.exc_info()
fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
typeE = str("TYPE : " + str(exc_type))
fileE = str("FILE : " + str(fname))
lineE = str("LINE : " + str(exc_tb.tb_lineno))
messageE = str("MESG : " + "\n" + str(errorMessage) + "\n")
print("\n" + typeE + 
          "\n" + fileE + 
          "\n" + lineE + 
          "\n" + messageE)
          `

---

# STRINGS & THINGS:

*Program to convert String to a List*

In [1]:
# Python code to convert string to list  
def Convert(string): #{
    li = list(string.split(" ")) 
    return li 
#}

In [2]:
# Driver code     
str1 = "Geeks for Geeks"
print(Convert(str1)) 

['Geeks', 'for', 'Geeks']


In [3]:
# Python code to convert string to list 
def Convert(string): #{
    li = list(string.split("-")) 
    return li 
#}

In [4]:
# Driver code     
str1 = "Geeks-for-Geeks"
print(Convert(str1)) 

['Geeks', 'for', 'Geeks']


## PRELIMINARY IMPORTS:

In [6]:
import pandas as pd
import numpy as np

In [29]:
s = pd.Series(['A', 'B', 'C', 'Aaba', 'Baca', np.nan, 'CABA', 'dog', 'cat'])

In [30]:
s

0       A
1       B
2       C
3    Aaba
4    Baca
5     NaN
6    CABA
7     dog
8     cat
dtype: object

In [31]:
s.str.lower()

0       a
1       b
2       c
3    aaba
4    baca
5     NaN
6    caba
7     dog
8     cat
dtype: object

# i) logging

In [1]:
import logging
logging.warning('Watch out!')  # will print a message to the console
logging.info('I told you so')  # will not print anything

In [ ]:
 # TRY THE FOLLOWING
try: #{
    logging.basicConfig(level=logging.INFO,
                        datefmt='%Y_%M_%d_%H%S%M',
                        filename='C:\data\outbound\Logs\CofA_renaming_etc.log',
                        filemode='a')
#}
except: #{
    errorMessage = str(sys.exc_info()[0]) + "\n"
    errorMessage = errorMessage + str(sys.exc_info()[1]) + "\n"
    errorMessage = errorMessage + str(sys.exc_info()[2]) + "\n"
    exc_type, exc_obj, exc_tb = sys.exc_info()
    fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
    print("TYPE : " + str(exc_type) + "\n")
    print("FILE : " + str(fname) + "\n")
    print("LINE-NO : " + str(exc_tb.tb_lineno) + "\n")
    print("MESSAGE : \n" + str(errorMessage))
#}
else: #{
    print("SUCCESS! VERY NICE!")
#}

# ii) cavaeats

`df[col] | Returns column with label col as Series`

`df[[col1, col2]] | Returns columns as a new DataFrame`

`s.iloc[0] | Selection by position`

`s.loc['index_one'] | Selection by index`

`df.iloc[0,:] | First row`

`df.iloc[0,0] | First element of first column`


# iii) OS-TOOLS:

In [2]:
import wmi, psutil

In [3]:
c = wmi.WMI()

In [4]:
for process in c.Win32_Process(): #{
    print(process.Name)
#}

System Idle Process
System
Secure System
smss.exe
csrss.exe
wininit.exe
csrss.exe
winlogon.exe
services.exe
LsaIso.exe
lsass.exe
svchost.exe
WUDFHost.exe
svchost.exe
fontdrvhost.exe
fontdrvhost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
dwm.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
TeraCopyService.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
NVDisplay.Container.exe
nvwmi64.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
Memory Compression
svchost.exe
svchost.exe
WUDFHost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
svchost.exe
hpservice.exe
svchost.exe
svchost.exe
vpnagent.exe
svchost.exe
svchost.exe
wlanext.exe
conhost.exe
spoolsv.exe
armsvc.exe
ciscod.exe
svchost.exe
AGMService.exe
enstart64.exe
AGSService.exe
C

In [ ]:
# TERMINATE EACH PROCESS
for process in c.Win32_Process(): #{
    if str(process.Name) == "opera.exe": #{
        print("FOUND OPERA!")
        opera_id = process.ProcessID
        print(opera_id)
        # create var representing process
        process_var = psutil.Process(opera_id)
        # KILL PROCESS
        process_var.terminate()
    #}
    else: #{
        print("DID NOT FIND OPERA!")
    #}
#}

---

# A) importing Data

## Excel

#### Preliminaries

In [32]:
import xlrd

In [33]:
inFile = "data/inbound/QCI&QCM Names.xlsx"

In [34]:
book = xlrd.open_workbook(str(inFile))

#### print out sheet names

In [35]:
for sheet in book.sheets():
    print(sheet.name)

Sheet1


#### define initial DataFrame

In [36]:
df = pd.read_excel(str(inFile), dtype=np.str, parse_dates=True, sheet_name=0, header=1, index_cols=0, usecols=1)

C:\Users\derbates\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\io\excel.py:520: UserWarning: The 'parse_dates=True' keyword of read_excel was provided without an 'index_col' keyword value.
  warn("The 'parse_dates=True' keyword of read_excel was provided"


In [37]:
df.head()

QCI-757                         QualityCheck Anions Sample
0  QCI-757A                QualityCheck Anions Sample - Part A
1  QCI-757B                QualityCheck Anions Sample - Part B
2   QCM-241  QualityCheck Aroclor 1242 In Transformer Oil S...
3   QCM-243  QualityCheck Aroclor 1248 In Transformer Oil S...
4   QCM-244  QualityCheck Aroclor 1248 In Transformer Oil S...

## 1) Time-Module(s), IO, logging?
---

### COUNTDOWN TIMER

#### Preliminaries

In [1]:
import time

In [2]:
def countdown(n): #{
    while n > 0: #{
        print (n)
        time.sleep (1 ) # sleep for 1 second
        n = n - 1
        if n == 0: #{
            print("BLAST OFF!")
        #}
    #}
#}

In [3]:
countdown(50)

50
49
48
47
46
45
44
43
42
41
40
39
38
37
36
35
34
33
32
31
30
29
28
27
26
25
24
23
22
21
20
19
18
17
16
15
14
13
12
11
10
9
8
7
6
5
4
3
2
1
BLAST OFF!


### Current Date and Time

In [1]:
import datetime
i = datetime.datetime.now()

In [2]:
print ("Current date & time = %s" % i)
print ("Date and time in ISO format = %s" % i.isoformat() )
print ("Current year = %s" %i.year)
print ("Current month = %s" %i.month)
print ("Current date (day) = %s" %i.day)
print ("dd/mm/yyyy format = %s/%s/%s" % (i.day, i.month, i.year) )
print ("Current hour = %s" %i.hour)
print ("Current minute = %s" %i.minute)
print ("Current second = %s" %i.second)
print ("hh:mm:ss format = %s:%s:%s" % (i.hour, i.month, i.second) )

Current date & time = 2019-05-03 09:10:33.906472
Date and time in ISO format = 2019-05-03T09:10:33.906472
Current year = 2019
Current month = 5
Current date (day) = 3
dd/mm/yyyy format = 3/5/2019
Current hour = 9
Current minute = 10
Current second = 33
hh:mm:ss format = 9:5:33


### Getting Loals Date and Time in Python

In [4]:
import time

now = time.strftime("%c")
## date and time representation
print("Current date & time " + time.strftime("%c"))
## Only date representation
print("Current date " + time.strftime("%x"))
## Only time representation
print("Current time " + time.strftime("%X"))
## Display current date and time from now variable
print(("Current time %s" % now ))

Current date & time Fri May  3 09:12:03 2019
Current date 05/03/19
Current time 09:12:03
Current time Fri May  3 09:12:03 2019


### Getting current time in pything using time module

The syntax is:

`time.strftime(format)`

`## 24 hour format ##`

`time.strftime("%H:%M:%S")`

`## 12 hour format ##`

`time.strftime("%I:%M:%S")`

In [5]:
#!/usr/anaconda3/bin/python

import time

print (time.strftime("%H:%M:%S"))
## 12 hour format ##
print (time.strftime("%I:%M:%S"))

09:14:08
09:14:08


### Creation-date of file(s) / INDEX OF CREATION DATES

#### Preliminaries

In [11]:
import os
import sys
from datetime import datetime

from pathlib import Path

In [2]:
inDirectory = "data/temp/zip/final/"

In [3]:
# create list of PDFS
pdf_list = os.listdir(inDirectory)

In [4]:
print("length of [pdf_list]: \n" + str((len(pdf_list))))

length of [pdf_list]: 
175


In [7]:
# Create Series off list
d1 = pd.Series(pdf_list)
d1.astype(dtype=np.str)

0           CofA-04-12-19-FINAL-2.zip
1             CofA_04-12-19_FINAL.zip
2               CUS-28911@CS-2462.PDF
3               CUS-28923@CS-2471.PDF
4               CUS-28951@CS-2579.PDF
5               CUS-28972@CS-2622.PDF
6               CUS-29034@CS-2713.PDF
7                EPA-1118@CP-0597.PDF
8                EPA-1169@CS-2012.PDF
9               EPA-1189@CL-0729A.PDF
10                NPM-108@CS-2034.PDF
11             NPM-108B-1@CH-0532.PDF
12               NPM-108B@CS-1114.PDF
13                 PP-420@CS-2759.PDF
14              PPS-642-1@CT-1302.pdf
15     PST-1010M100A01@0006453617.pdf
16          PST-1025A1000@CS-1733.PDF
17          PST-1025A1000@CT-0937.pdf
18          PST-1035A1000@CT-1054.pdf
19       PST-1055A1000@0006453618.pdf
20          PST-1065A1000@CT-0563.pdf
21         PST-1205A1000@CP-3522Z.PDF
22        PST-1220M100A01@CT-0906.pdf
23          PST-1315A1000@CT-0934.pdf
24        PST-1355M100A01@CT-0943.pdf
25        PST-1440M100A01@CT-0942.pdf
26          

In [17]:
dfDateList = pd.DataFrame()

In [18]:
dfDateList['PDFizzle'] = d1

In [22]:


try:
    # Create a column var to store the data
    dateCol = []
    
    # counter
    x = 0
    
    # create Series from column
    d1 = pd.Series(dfDateList['PDFizzle'])
    d1.astype(dtype=np.str)
    
    # for each row in the column
    for row in d1:
        theStr = str(row)
        testPath = Path(inDirectory + theStr)
        # create date var
        date = os.path.getctime(testPath)
        # FORMAT date var & SET AS str
        dateStr = str(datetime.fromtimestamp(date))
        print("CREATE-DATE == " + dateStr)
        dateCol.append(dateStr)
        x = x + 1

    # Create column from list
    dfDateList['Create-Date'] = dateCol
except:
    errorMessage = str(sys.exc_info()[0]) + "n\t\t"
    errorMessage = errorMessage + str(sys.exc_info()[1]) + "\n\t\t"
    errorMessage = errorMessage + str(sys.exc_info()[2]) + "\n"
    exc_type, exc_obj, exc_tb = sys.exc_info()
    fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
    typeE = str("TYPE : " + str(exc_type))
    fileE = str("FILE : " + str(fname))
    lineE = str("LINE : " + str(exc_tb.tb_lineno))
    messageE = str("MESG : " + "\n\n" + str(errorMessage) + "\n")
    print("\n" + typeE + 
          "\n" + fileE + 
          "\n" + lineE + 
          "\n" + messageE)
else:
    print("FIN...")

CREATE-DATE == 2019-04-12 16:02:47.028306
CREATE-DATE == 2019-04-12 15:17:21.952837
CREATE-DATE == 2019-04-12 11:11:38.740778
CREATE-DATE == 2019-04-12 11:11:16.307054
CREATE-DATE == 2019-04-12 12:09:28.931834
CREATE-DATE == 2019-04-12 11:37:13.786123
CREATE-DATE == 2019-04-12 11:23:18.768490
CREATE-DATE == 2019-04-11 17:13:28.860189
CREATE-DATE == 2019-04-11 17:14:00.879953
CREATE-DATE == 2019-04-11 17:14:15.336493
CREATE-DATE == 2019-04-11 17:39:03.420662
CREATE-DATE == 2019-04-12 11:38:55.147079
CREATE-DATE == 2019-04-11 17:39:04.809097
CREATE-DATE == 2019-04-11 17:42:39.884207
CREATE-DATE == 2019-04-12 11:59:34.914365
CREATE-DATE == 2019-04-12 11:55:22.563747
CREATE-DATE == 2019-04-12 12:09:41.152809
CREATE-DATE == 2019-04-12 12:09:41.121807
CREATE-DATE == 2019-04-12 11:35:00.343303
CREATE-DATE == 2019-04-12 11:36:03.123747
CREATE-DATE == 2019-04-12 11:41:39.134271
CREATE-DATE == 2019-04-12 11:53:53.777912
CREATE-DATE == 2019-04-12 11:13:19.078440
CREATE-DATE == 2019-04-12 11:56:16

In [24]:
dfDateList.index = dfDateList['Create-Date']

In [25]:
dfDateList.head(25)

PDFizzle  \
Create-Date                                                  
2019-04-12 16:02:47.028306       CofA-04-12-19-FINAL-2.zip   
2019-04-12 15:17:21.952837         CofA_04-12-19_FINAL.zip   
2019-04-12 11:11:38.740778           CUS-28911@CS-2462.PDF   
2019-04-12 11:11:16.307054           CUS-28923@CS-2471.PDF   
2019-04-12 12:09:28.931834           CUS-28951@CS-2579.PDF   
2019-04-12 11:37:13.786123           CUS-28972@CS-2622.PDF   
2019-04-12 11:23:18.768490           CUS-29034@CS-2713.PDF   
2019-04-11 17:13:28.860189            EPA-1118@CP-0597.PDF   
2019-04-11 17:14:00.879953            EPA-1169@CS-2012.PDF   
2019-04-11 17:14:15.336493           EPA-1189@CL-0729A.PDF   
2019-04-11 17:39:03.420662             NPM-108@CS-2034.PDF   
2019-04-12 11:38:55.147079          NPM-108B-1@CH-0532.PDF   
2019-04-11 17:39:04.809097            NPM-108B@CS-1114.PDF   
2019-04-11 17:42:39.884207              PP-420@CS-2759.PDF   
2019-04-12 11:59:34.914365           PPS-642-1@CT-1302.pdf   
2019-04-12 11:55:22.563747  PST-1010M100A01@0006453617.pdf   
2019-04-12 12:09:41.152809       PST-1025A1000@CS-1733.PDF   
2019-04-12 12:09:41.121807       PST-1025A1000@CT-0937.pdf   
2019-04-12 11:35:00.343303       PST-1035A1000@CT-1054.pdf   
2019-04-12 11:36:03.123747    PST-1055A1000@0006453618.pdf   
2019-04-12 11:41:39.134271       PST-1065A1000@CT-0563.pdf   
2019-04-12 11:53:53.777912      PST-1205A1000@CP-3522Z.PDF   
2019-04-12 11:13:19.078440     PST-1220M100A01@CT-0906.pdf   
2019-04-12 11:56:16.378677       PST-1315A1000@CT-0934.pdf   
2019-04-12 11:36:16.070299     PST-1355M100A01@CT-0943.pdf   

                                           Create-Date  
Create-Date                                             
2019-04-12 16:02:47.028306  2019-04-12 16:02:47.028306  
2019-04-12 15:17:21.952837  2019-04-12 15:17:21.952837  
2019-04-12 11:11:38.740778  2019-04-12 11:11:38.740778  
2019-04-12 11:11:16.307054  2019-04-12 11:11:16.307054  
2019-04-12 12:09:28.931834  2019-04-12 12:09:28.931834  
2019-04-12 11:37:13.786123  2019-04-12 11:37:13.786123  
2019-04-12 11:23:18.768490  2019-04-12 11:23:18.768490  
2019-04-11 17:13:28.860189  2019-04-11 17:13:28.860189  
2019-04-11 17:14:00.879953  2019-04-11 17:14:00.879953  
2019-04-11 17:14:15.336493  2019-04-11 17:14:15.336493  
2019-04-11 17:39:03.420662  2019-04-11 17:39:03.420662  
2019-04-12 11:38:55.147079  2019-04-12 11:38:55.147079  
2019-04-11 17:39:04.809097  2019-04-11 17:39:04.809097  
2019-04-11 17:42:39.884207  2019-04-11 17:42:39.884207  
2019-04-12 11:59:34.914365  2019-04-12 11:59:34.914365  
2019-04-12 11:55:22.563747  2019-04-12 11:55:22.563747  
2019-04-12 12:09:41.152809  2019-04-12 12:09:41.152809  
2019-04-12 12:09:41.121807  2019-04-12 12:09:41.121807  
2019-04-12 11:35:00.343303  2019-04-12 11:35:00.343303  
2019-04-12 11:36:03.123747  2019-04-12 11:36:03.123747  
2019-04-12 11:41:39.134271  2019-04-12 11:41:39.134271  
2019-04-12 11:53:53.777912  2019-04-12 11:53:53.777912  
2019-04-12 11:13:19.078440  2019-04-12 11:13:19.078440  
2019-04-12 11:56:16.378677  2019-04-12 11:56:16.378677  
2019-04-12 11:36:16.070299  2019-04-12 11:36:16.070299

### Opening a 'file-like' document

#### Preliminaries

In [1]:
from docx import Document
from docx.shared import Inches
from docx.shared import Pt

python-docx can open a document from a so-called file-like object. It can also save to a file-like object. This can be handy when you want to get the source or target document over a network connection or from a database and don’t want to (or aren’t allowed to) interact with the file system. In practice this means you can pass an open file or StringIO/BytesIO stream object to open or save a document like so:

In [ ]:
f = open('foobar.docx', 'rb')
document = Document(f)
f.close()

# or

with open('foobar.docx', 'rb') as f:
    source_stream = StringIO(f.read())
document = Document(source_stream)
source_stream.close()
...
target_stream = StringIO()
document.save(target_stream)

The 'rb' file open mode parameter isn’t required on all operating systems. It defaults to 'r' which is enough sometimes, but the ‘b’ (selecting binary mode) is required on Windows and at least some versions of Linux to allow Zipfile to open the file.

Okay, so you’ve got a document open and are pretty sure you can save it somewhere later. Next step is to get some content in there …

In [38]:
# import time
import time; # This is required to include time module.

ticks = time.time()
print(str("Number of ticks since 12:00am, January 1, 1970:"), ticks)

Number of ticks since 12:00am, January 1, 1970: 1552681870.2749608


In [39]:
localtime = time.asctime( time.localtime(time.time()) )
print(str("Local current time :"), localtime)

Local current time : Fri Mar 15 16:31:10 2019


In [40]:
# import DATE-TIME
## used for file-convention_naming
import datetime

# CREATE TIME STAMP FOR FILE CREATION NAME
date_object = datetime.date.today()
#hour_object = datetime.datetime.hour()
#min_object = datetime.datetime.minute()
#sec_object = datetime.datetime.second()
time_object = str('{0:%m_%d_%Y-%H%M%S}'.format(datetime.datetime.now())) # _%S <--- FOR SECONDS
print(str("file naming convention (today's example) :"), time_object)

file naming convention (today's example) : 03_15_2019-163110


In [41]:
matchFile = "data/outbound/FILE-NAME-" + str(time_object) + ".csv"

In [42]:
print(matchFile)

data/outbound/FILE-NAME-03_15_2019-163110.csv


### displaying style types for Document

In [1]:
from docx import Document
from docx.shared import Inches
from docx.shared import Pt

from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.style import WD_STYLE_TYPE

################################################

import os


In [2]:
def displayStyles(theDoc):
    
    document = Document(theDoc)
    
    styles = document.styles
    paragraph_styles = [
        s for s in styles if s.type == WD_STYLE_TYPE.PARAGRAPH
    ]
    for style in paragraph_styles:
        print(style.name)

In [3]:
inDirectory = "data/inbound/QCI&QCM Word Files Rev 4.0_legacy/"

In [6]:
# Create list of String
doc_list = os.listdir(inDirectory)

In [8]:
testDoc = str(doc_list[4])

In [9]:
displayStyles(inDirectory + testDoc)

Normal
Header
Footer
Balloon Text


### outputting METHODS of any function/module/etc

In [3]:
dir(Document)

['__annotations__',
 '__call__',
 '__class__',
 '__closure__',
 '__code__',
 '__defaults__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__get__',
 '__getattribute__',
 '__globals__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__kwdefaults__',
 '__le__',
 '__lt__',
 '__module__',
 '__name__',
 '__ne__',
 '__new__',
 '__qualname__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__']

In [5]:
dir(os.system)

['__call__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__name__',
 '__ne__',
 '__new__',
 '__qualname__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__self__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__text_signature__']

## 2) Filtering, sorting, and searching
---

### REGULAR EXPRESSIONS:

In [11]:
import re

In [12]:
line = "Cats are smarter than dogs";

In [13]:
matchObj = re.match( r'dogs', line, re.M|re.I)

In [14]:
if matchObj:
    print("Match --> matchObj.group() : ", matchObj.group())
else:
    print("no match!!")

no match!!


In [15]:
searchObj = re.search( r'dogs', line, re.M|re.I)

In [17]:
if searchObj:
    print("search --> searchObj.group() : ", searchObj.group())
else:
    print("Nothing found!!")

search --> searchObj.group() :  dogs


### Mass renaming of Columns:
`df.rename(columns=lambda x: x + 1)`

### Return the sum of the values for the requested axis
`pd.Series([]).sum()  # min_count=0 is the default`

### Convert column of DataFrame from string to float
`dfMachies = pd.DataFrame()
 cols = ['umeasur', 'umfact', 'qtyord']                         
 for col in cols:                                               
     dfMatchies[col] = dfMatchies[col].astype(dtype=np.float64)`

## 3) Mathmatical Operations & loops
---

**IN ORDER TO SKIP DIRECTORIES**:

In [ ]:
import os, sys, time
from time import sleep
from pathlib import Path

In [ ]:
# create list to hold clean data yo
path_list = []

# counter
x = 0

for root, dirs, files in os.walk(in_directory): #{
    # DIRECTORY SKIP CONDITIONS
    if 'Archive ERR' in dirs: #{
        dirs.remove('Archive ERR')  # don't visit Archive ERR directories
    #}
    if 'Archive - For all archive CofA, see G CofA folder' in dirs: #{
        dirs.remove('Archive - For all archive CofA, see G CofA folder') # skip
    #}
    if 'Instruction Sheets' in dirs: #{
        dirs.remove('Instruction Sheets') # skip
    #}
    # OTHERWISE IF THERE ARE FILES IN DIRECTORY
    for f in files: #{
        # IF FILE IS OF TYPE .pdf
        if fnmatch.fnmatch(f, "*.pdf"): #{
            # ASSEMBLE!
            file_path = os.path.join(root, f)
            """
            # GET CREATE TIME
            create_time = os.path.getctime(file_path)
            # GET MODIFIED TIME
            mod_time = os.path.getmtime(file_path)
            # MAKE 'ctime' INTO DATETIME / more readable
            readable_c = datetime.datetime.fromtimestamp(create_time).isoformat()
            # MAKE 'mtime' INTO DATETIME / more readable
            readable_m = datetime.datetime.fromtimestamp(mod_time).isoformat()
            # CONVERT TO pandas.Timestamp
            timeStamp_c = pd.Timestamp(readable_c)
            # AGAIN FOR MODIFIED TIME
            timeStamp_m = pd.Timestamp(readable_m)
            """
            path_list.append(file_path)
            x += 1
        #}
        else: #{
            print("NOT A PDF : \t" + str(f))
        #}
    #}
#}

In [ ]:
# CREATE NEW DATAFRAME TO HOLD LIST
df_paths = pd.DataFrame()
# ASSIGN LIST TO COLUMN IN DATAFRAME
df_paths['CofA File'] = path_list
df_paths.to_csv("C:\CofA\log\CofA_Email_Node_List_" + time_now + "_pull.csv", index=False)
print("\n COUNT == " + str(x))

iterate thru a directory of files and print details for each **SEPERATE**

 1. root
 2. directories
 3. files
 
#### truncates through tree(s) until everything underneath is found 

### 3a) loop through and print:
- root
- dict(s)
- file(s)

#### Preliminaries

In [2]:
import os
from pathlib import Path

In [3]:
inDirectory = Path("data/outbound/")

In [9]:
# iterate thru directory/directories 
for root,d_names,f_names in os.walk(inDirectory):
    print("===================== START ============================================== \n \n")
    print("\t *root* = " + str(root) + "\n") 
    print("\t *dict* = " + str(d_names) + "\n")
    if len(f_names) >= 0:
        print("\t *file(s)*: \n")
        # multiple file(s)
        for f in f_names:
            print ("\t " + str(f) + "\n")
    else:
        # no files
        print("\t NO-FILES!")
    print("\n \n================================================ END ===================== \n")

===================== START ============================================== 
 

	 *root* = data\outbound

	 *dict* = ['archive', 'project 2', 'QCI_QCM_complete', '_log', '_test']

	 *file(s)*: 

	 Clean_Container_List_fixed.csv

	 dfMatchies_Dropped_03_14_2019-100303.csv

	 dfMatchies_purno_Dropped_03_14_2019-100303.csv

	 dfMatchies_purno_Dropped_03_14_2019-100815.csv

	 dfMatchies_purno_Dropped_03_14_2019-100918.csv

	 dfMatchies_purno_Dropped_03_14_2019-101219.csv

	 dfMatchies_purno_Dropped_03_14_2019-105236.csv

	 dfMatchies_purno_Dropped_03_14_2019-111415.csv

	 dfMatchies_purno_Dropped_03_14_2019-114046.csv

	 dfMatchies_purno_Dropped_03_14_2019-114231.csv

	 dfMatchies_purno_Dropped_03_14_2019-115808.csv

	 dfMatchies_purno_Dropped_03_14_2019-120550.csv

	 dfMatchies_purno_Dropped_03_14_2019-122428.csv

	 dfMatchies_purno_Dropped_03_14_2019-124444.csv

	 dfMatchies_purno_Dropped_03_14_2019-124534.csv

	 dfMatchies_purno_Dropped_03_14_2019-130212.csv

	 dfMatchies_purno_Dropped_0

	 03_21_2019-111114_TEST_2_LIST.csv

	 03_21_2019-111114_TEST_3_LIST.csv

	 03_21_2019-111114_TEST_LIST.csv

	 03_21_2019-111207.csv

	 03_21_2019-111207_DOC_LIST.csv

	 03_21_2019-111207_PROD_LIST.csv

	 03_21_2019-111207_test.docx

	 03_21_2019-111207_TEST_2_LIST.csv

	 03_21_2019-111207_TEST_3_LIST.csv

	 03_21_2019-111207_TEST_LIST.csv

	 03_21_2019-111931.csv

	 03_21_2019-111931_DOC_LIST.csv

	 03_21_2019-111931_PROD_LIST.csv

	 03_21_2019-111931_test.docx

	 03_21_2019-111931_TEST_2_LIST.csv

	 03_21_2019-111931_TEST_3_LIST.csv

	 03_21_2019-111931_TEST_LIST.csv

	 03_21_2019-113725_PROD_LIST.csv

	 03_21_2019-113725_test.docx

	 03_21_2019-113812_PROD_LIST.csv

	 03_21_2019-113812_test.docx

	 03_21_2019-113847_PROD_LIST.csv

	 03_21_2019-113847_test.docx

	 03_21_2019-113930_PROD_LIST.csv

	 03_21_2019-113930_test.docx

	 03_21_2019-114031_PROD_LIST.csv

	 03_21_2019-114031_test.docx

	 03_21_2019-114438_PROD_LIST.csv

	 03_21_2019-114438_test.docx

	 03_21_2019-114906_PROD_LI

`
from zipfile_infolist import print_info
import zipfile
print 'creating archive'
zf = zipfile.ZipFile('zipfile_write.zip', mode='w')
try:
print 'adding README.txt'
zf.write('README.txt')
finally:
print 'closing'
zf.close()
print
print_info('zipfile_write.zip')
`

### 3b) loop through directory of files and print each filename

#### Preliminaries

In [43]:
import os
inDirectory = "data/inbound/QCI&QCM Word Files Rev 4.0/"

#### use **os.listdir** func

In [ ]:
os.listdir(inDirectory)

### 3c) loop through and print file names (different way)

In [ ]:
destdir = Path(inDirectory)
try: #{
    # counter
    i = 0
    files = [p for p in destdir.iterdir() if p.is_file()]
    for p in files: #{
        print("<BEGIN LOOP # " + str(i) + "> ======================<> \n| =")
        with p.open() as f: #{
            theStr = str(p)
            theLen = len(theStr)
            print("| PATH : " + theStr)
        #}
    #}
#}

### 4a) Create A pandas Column With A For Loop

#### Preliminaries

In [45]:
import pandas as pd
import numpy as np

#### Create an example dataframe

In [46]:
raw_data = {'student_name': ['Miller', 'Jacobson', 'Ali', 'Milner', 'Cooze', 'Jacon', 'Ryaner', 'Sone', 'Sloan', 'Piger', 'Riani', 'Ali'], 
        'test_score': [76, 88, 84, 67, 53, 96, 64, 91, 77, 73, 52, np.NaN]}
df = pd.DataFrame(raw_data, columns = ['student_name', 'test_score'])

In [47]:
# View that FRESH frame
df

student_name  test_score
0        Miller        76.0
1      Jacobson        88.0
2           Ali        84.0
3        Milner        67.0
4         Cooze        53.0
5         Jacon        96.0
6        Ryaner        64.0
7          Sone        91.0
8         Sloan        77.0
9         Piger        73.0
10        Riani        52.0
11          Ali         NaN

#### Create a function to assign letter grades

In [48]:
# Create a list to store the data
grades = []

# For each row in the column,
for row in df['test_score']:
    # if more than a value,
    if row > 95:
        # Append a letter grade
        grades.append('A')
    # else, if more than a value,
    elif row > 90:
        # Append a letter grade
        grades.append('A-')
    # else, if more than a value,
    elif row > 85:
        # Append a letter grade
        grades.append('B')
    # else, if more than a value,
    elif row > 80:
        # Append a letter grade
        grades.append('B-')
    # else, if more than a value,
    elif row > 75:
        # Append a letter grade
        grades.append('C')
    # else, if more than a value,
    elif row > 70:
        # Append a letter grade
        grades.append('C-')
    # else, if more than a value,
    elif row > 65:
        # Append a letter grade
        grades.append('D')
    # else, if more than a value,
    elif row > 60:
        # Append a letter grade
        grades.append('D-')
    # otherwise,
    else:
        # Append a failing grade
        grades.append('Failed')
        
# Create a column from the list
df['grades'] = grades

In [49]:
# View the new dataframe
df

student_name  test_score  grades
0        Miller        76.0       C
1      Jacobson        88.0       B
2           Ali        84.0      B-
3        Milner        67.0       D
4         Cooze        53.0  Failed
5         Jacon        96.0       A
6        Ryaner        64.0      D-
7          Sone        91.0      A-
8         Sloan        77.0       C
9         Piger        73.0      C-
10        Riani        52.0  Failed
11          Ali         NaN  Failed

### 3b) looping thru rows of single column to peform STRING ops

#### Create a function to change string names/values

In [50]:
# Create list to store data
qtyCol = []

"g = []"
"mL = []"

# create a Series from 'umeasur' column 
q1 = pd.Series(dfFinalFrame['umeasur'])
q1.astype(dtype=np.str)

# for each row in column
for row in q1:
    theStr = str(row)
    rowLen = len(theStr)
    
     # if we --DO-- find a match with G
    if int(theStr.find('G', 0, len(str(row)))) != -1:
        print("FOUND: at symbol == " + str(theStr))
        aStr = str("G")
        print(aStr)
    # else, if we --DO-- find a match with KG
    elif int(theStr.find('KG', 0, rowLen)) != -1:
        print("FOUND: at symbol == " + str(theStr))
        aStr = str("G")
        print(aStr)
    # else, if we --DO-- find a match with MG
    elif int(theStr.find('MG', 0, rowLen)) != -1:
        print("FOUND: at symbol == " + str(theStr))
        aStr = str('G')
        print(aStr)
    # else, if we --DO-- find a match with mL
    elif int(theStr.find('mL', 0, len(str(row)))) != -1:
        print("FOUND: at symbol == " + str(theStr))
        aStr = str("mL")
        print(aStr)
    # else, if we --DO-- find a match with ML
    elif int(theStr.find('ML', 0, rowLen)) != -1:
        print("FOUND: at symbol == " + str(theStr))
        aStr = str('mL')
        print(aStr)
    # else, WE DID NOT FIND ANY OF THE ABOVE MATCHIES
    else:
        aStr = str('N/A')
        print(aStr)
    qtyCol.append(aStr)
    
    

    
# Create a column from the list
dfFinalFrame['qty'] = qtyCol


NameError: name 'dfFinalFrame' is not defined

## 4) Cleanup, Renaming, Joining, and Scrubbing

### 4a) renaming index column of a DataFrame

#### You can just get/set the index via its `name` property

In [ ]:
df.index.name

In [ ]:
df.index.name = 'foo'

In [ ]:
df.index.name

In [ ]:
df

### 4b) Creating a MultiIndex (hierarchical index) object

The **`MultiIndex`** object is the hierarchical analogue of the standard **`Index`** object which typically stores the axis labels in pandas objects.

You can think of `MultiIndex` as an array of tuples where each tuple is unique. 

A `MultiIndex` can be created from a list of arrays (using **`MultiIndex.from_arrays()`**), an array of tuples (using **`MultiIndex.from_tuples()`**), a crossed set of iterables (using **`MultiIndex.from_product()`**), or a DataFrame (using **`MultiIndex.from_frame()`**). 

The Index constructor will attempt to return a `MultiIndex` when it is passed a list of tuples. The following examples demonstrate different ways to initialize MultiIndexes.

In [1]:
arrays = [['bar', 'bar', 'baz', 'baz', 'foo', 'foo', 'qux', 'qux'],
          ['one', 'two', 'one', 'two', 'one', 'two', 'one', 'two']]

In [2]:
arrays

[['bar', 'bar', 'baz', 'baz', 'foo', 'foo', 'qux', 'qux'],
 ['one', 'two', 'one', 'two', 'one', 'two', 'one', 'two']]

In [3]:
tuples = list(zip(*arrays))

In [4]:
tuples

[('bar', 'one'),
 ('bar', 'two'),
 ('baz', 'one'),
 ('baz', 'two'),
 ('foo', 'one'),
 ('foo', 'two'),
 ('qux', 'one'),
 ('qux', 'two')]

In [7]:
index = pd.MultiIndex.from_tuples(tuples, names=['first', 'second'])

In [8]:
index

MultiIndex(levels=[['bar', 'baz', 'foo', 'qux'], ['one', 'two']],
           labels=[[0, 0, 1, 1, 2, 2, 3, 3], [0, 1, 0, 1, 0, 1, 0, 1]],
           names=['first', 'second'])

In [9]:
s = pd.Series(np.random.randn(8), index=index)

In [10]:
s

first  second
bar    one      -0.438025
       two      -1.049278
baz    one       0.611275
       two      -0.842298
foo    one      -1.288066
       two      -0.683080
qux    one       0.472040
       two       1.681105
dtype: float64

### 4c) Joining two DataFrames (identicial cols)

#### Add the rows in df1 to the end of df2 (columns should be identical)

In [ ]:
df1.append(df2)

#### Add the columns in df1 to the end of df2 (rows should be identical)

In [ ]:
pd.concat([df1, df2],axis=1)

---

### *ENCODING*

*ISO-8859-1*


---

`
dfCofaList_2 = pd.read_csv(str(inFile_2), encoding="ISO-8859-1")
`

---

## 5) Error-Code Index: Raises & Solutions

### 5a) error codes

### *UnicodeDocodeError*: 

'utf-8' codec can't decode byte 0xb5 in position 5: invalid start byte

## 6) ZIP files, PDFs, docx, xlsx, txt, ETC

#### 6A) Reading Meta-data from a ZIP Archive

##### 6B) Extracting Files from ZIP Aarchive

#### 6C) Creating New Archives (ZIPS)

To Create a new ZIP archive, you open a ZipFile object in write mode `(w)` and add the files you want to archive:

In [ ]:
import zipfile

file_list = ['file1.py', 'sub_dir/', 'sub_dir/bar.py', 'sub_dir/foo.py']

with zipfile.ZipFile('new.zip', 'w') as new_zip:
    for name in file_list:
        new_zip.write(name)

In the example, `new_zip` is opened in write mode and each file in `file_list` is added to the archive. 

When the `with` statement suite is finished, `new_zip` is closed. 

Opening a ZIP file in write mode erases the contents of the archive and creates a new archive.

#### 6D) Add Files to *EXISTING* Archive (ZIP)

To add files to an existing archive, open a `ZipFile` object in append mode and then add the files:

In [ ]:
# Open a ZipFile object in append mode
with zipfile.ZipFile('new.zip', 'a') as new_zip:
    new_zip.write('data.txt')
    new_zip.write('latin.txt')

Here, you open the `new.zip` archive you created in the previous example in append mode. 

Opening the `ZipFile` object in append mode allows you to add new files to the ZIP file without deleting its current contents. 

After adding files to the ZIP file, the `with` statement goes out of context and closes the ZIP file.

---
## 7) Copy, Move, Create *FILES*

### `shutil.copy(src, dst, *, follow_symlinks=True)`

Copies the file src to the file or directory dst. src and dst should be strings. If dst specifies a directory, the file will be copied into dst using the base filename from src. Returns the path to the newly created file.

If *follow_symlinks* is false, and src is a symbolic link, dst will be created as a symbolic link. If *follow_symlinks* is true and src is a symbolic link, dst will be a copy of the file src refers to.

`copy()` copies the file data and the file’s permission mode (see `os.chmod()`). Other metadata, like the file’s creation and modification times, is not preserved. To preserve all file metadata from the original, use `copy2()` instead.

## `shutil.copyfile(src, dst, *, follow_symlinks=True)`

Copy the contents (no metadata) of the file named src to a file named dst and return dst. src and dst are path names given as strings. dst must be the complete target file name; look at shutil.copy() for a copy that accepts a target directory path. If src and dst specify the same file, SameFileError is raised.

The destination location must be writable; otherwise, an OSError exception will be raised. If dst already exists, it will be replaced. Special files such as character or block devices and pipes cannot be copied with this function.

If follow_symlinks is false and src is a symbolic link, a new symbolic link will be created instead of copying the file src points to.

Changed in version 3.3: IOError used to be raised instead of OSError. Added follow_symlinks argument. Now returns dst.

Changed in version 3.4: Raise SameFileError instead of Error. Since the former is a subclass of the latter, this change is backward compatible.

## 8) Archive & SAVING

## `shutil.make_archive(base_name, format[, root_dir[, base_dir[, verbose[, dry_run[, owner[, group[, logger]]]]]]])`

Create an archive file (such as zip or tar) and return its name.

base_name is the name of the file to create, including the path, minus any format-specific extension. format is the archive format: one of “zip” (if the zlib module is available), “tar”, “gztar” (if the zlib module is available), “bztar” (if the bz2 module is available), or “xztar” (if the lzma module is available).

root_dir is a directory that will be the root directory of the archive; for example, we typically chdir into root_dir before creating the archive.

base_dir is the directory where we start archiving from; i.e. base_dir will be the common prefix of all files and directories in the archive.

root_dir and base_dir both default to the current directory.

If dry_run is true, no archive is created, but the operations that would be executed are logged to logger.

owner and group are used when creating a tar archive. By default, uses the current owner and group.

logger must be an object compatible with PEP 282, usually an instance of logging.Logger.

The verbose argument is unused and deprecated.